go/vt/vtgate/planbuilder/dml.go

```diff
@@ -87,30 +106,51 @@ func isIndexChanging(setClauses sqlparser.UpdateExprs, colVindexes []*vindexes.C
 
 // buildUpdatePlan builds the instructions for a DELETE statement.
 func buildDeletePlan(del *sqlparser.Delete, vschema VSchema) (*engine.Route, error) {
-	route := &engine.Route{
+	er := &engine.Route{
 		Query: generateQuery(del),
 	}
+	aliased, ok := del.TableExprs[0].(*sqlparser.AliasedTableExpr)
+	if !ok || del.Targets != nil || len(del.TableExprs) != 1 {
+		bldr, err := processTableExprs(del.TableExprs, vschema)
+		if err != nil {
+			return nil, err
+		}
+		rb, ok := bldr.(*route)
+		if !ok || rb.ERoute.Keyspace.Sharded {
+			return nil, errors.New("unsupported: multi-table delete statement in sharded keyspace")
+		}
+		if hasSubquery(del) {
+			return nil, errors.New("unsupported: subqueries in DML")
+		}
+		er.Keyspace = rb.ERoute.Keyspace
+		er.Opcode = engine.DeleteUnsharded
+		return er, nil
+	}
+	updateTable, ok := aliased.Expr.(sqlparser.TableName)
```

Comment:
We could follow the same logic as update here. It's also beneficial to keep the two flows similar (for readability). Instead, we can make some tweaks below:
* Everything up to the point where we check for unsharded keyspace remains similar to update.
* If the keyspace is sharded,  then we check `del.Targets != nil || er.Table == nil`. In other words, a unique table was not identified. So, we return an error.

I think this approach will work. Let me know what you think.